File: system-design/oop-patterns/04-factory-method-pattern/full/java/zoo/RandomAnimalFactory.java
package zoo;

public class RandomAnimalFactory implements AnimalFactory {
    public Animal createAnimal() {
        int choice = new java.util.Random().nextInt(3);
        if (choice == 0) return new Dog();
        if (choice == 1) return new Cat();
        return new Duck();
    }
}

File: system-design/oop-patterns/04-factory-method-pattern/full/ts/zoo/RandomAnimalFactory.ts
import type { Animal } from './Animal';
import type { AnimalFactory } from './AnimalFactory';
import { Dog } from './Dog';
import { Cat } from './Cat';
import { Duck } from './Duck';

export class RandomAnimalFactory implements AnimalFactory {
  createAnimal(): Animal {
    const choice = Math.floor(Math.random() * 3);
    if (choice === 0) return new Dog();
    if (choice === 1) return new Cat();
    return new Duck();
  }
}
